### ISSUE TITLE
inconsistency in undefined function argument as not optional

### ISSUE THREAD

**Issue by npenin** [microsoft/TypeScript#35350]
**TypeScript Version:**  3.7.2

**Search Terms:** undefined optional function argument 

**Code**

```ts
function f(a?: boolean)
{
    return !!a;
}

declare type Argument0<T> = T extends ((x: infer X, ...z: any[]) => any) ? X : never;

function g(a: Argument0<typeof f>)
{
    return !!a;
}

function h(a: boolean | undefined)
{
    return !!a;
}

f();
g();
h();
```

**Expected behavior:** g and h should be considered valid as tooltip on a in f, g and h are reporting parameter a is `boolean | undefined`

**Actual behavior:** g and h are considered invalid calls.

**Playground Link:** http://www.typescriptlang.org/play/index.html#code/GYVwdgxgLglg9mABMAFAQwPwC5ECM5wA2ApmmAJQBQA3pYvYgE7FQiNICEHaA3JQL6VKAE2IRCaZoigBPAA7FEAQUYBzEAFtiYKAAYAPABUAfIgC8iQ4mIAPKNuEBnRChQ2cMMMGKNEADQAaRAA6UIAvHDIZAG0AXXJzUyiEjH9EHDBiADcfPkpQSFgERFV0HBV1LR0DWQU4YGRjKloGJhY2Tm4+QXzwaHgkAAsyvAISMkQAH0RwUWBPYmFmugZmVnZELl4BIVRyPlL9ymGjoA

**Related Issues:** 

**EDIT** swaped expected and actual behavior

**Comment by MartinJohns** [comment]
> Expected behavior: g and h are considered invalid calls.

`g` and `h` **are** considered invalid in the playground link you provided. So... everything working as expected?

**Comment by paulsouche** [comment]
`boolean | undefined` doesn't mean optional. `g(undefined); and h(undefined);` or
```
function g(a?: Argument0<typeof f>)
{
    return !!a;
}
g(); // this is valid
```

are valid calls

**Comment by npenin** [comment]
sorry, I mixed expected and actual behavior. Well there is a discrepancy between `a?: boolean` which translates to `boolean | undefined` and having `a:boolean|undefined`. When calling a function, `undefined` is assumed for all the unspecified parameters, which makes them optional. This is exactly how you test whether a parameter was given in the call to a function `typeof a=='undefined'`.

**Comment by VulcanRav** [comment]
`const foo: {test?: string} = {test: undefined} // okay (shouldn't imho)`
`const bar: {test: string | undefined } = foo // fails`

I think {test: undefined} should not be assignable to {test?: string}.

The meaning of {test?: string} should be: if there is the key test it's a string.

This is inconsistent behavior indeed.

**Comment by MartinJohns** [comment]
> The meaning of {test?: string} should be: if there is the key test it's a string.

But it's not. It means the property is optional, and it may be either `string` or `undefined`.

Such behavior is currently not possible and would require #13195 (distinguish missing and undefined).

**Comment by VulcanRav** [comment]
I agree, which is why I say it is inconsistent, in my previous example `foo` should be assignable to `bar` then, but it's not because bar **is** distinguishing between a missing and an undefined.

**Comment by VulcanRav** [comment]
It gets actually even worse, it seems that I can't find a way to create a truly `optional` type.

How can this be okay in typescript?

```typescript
type A = {} | { a: number };
const t: A = { a: undefined };
```

**Comment by RyanCavanaugh** [comment]
I don't have a lot to add here because there's a lot of basic misunderstanding happening. This kind of discussion is best suited for StackOverflow or other venues; none of the behavior described is a bug or other defect.

**Comment by npenin** [comment]
@RyanCavanaugh I don't see the misunderstanding in my original request, especially with the inferred type. I still believe that `undefined` is/should be the equivalent of optional as this is how Javascript behaves. the `?` is just syntaxic sugar to help having more readable code.

**Comment by RyanCavanaugh** [comment]
We made different syntax do different things so that you could pick which thing you wanted to happen.
